opentelemetry/proto/profiles/v1experimental/pprofextended.proto

```diff
@@ -0,0 +1,386 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Profile is a common stacktrace profile format.
+//
+// Measurements represented with this format should follow the
+// following conventions:
+//
+// - Consumers should treat unset optional fields as if they had been
+//   set with their default value.
+//
+// - When possible, measurements should be stored in "unsampled" form
+//   that is most useful to humans.  There should be enough
+//   information present to determine the original sampled values.
+//
+// - On-disk, the serialized proto must be gzip-compressed.
+//
+// - The profile is represented as a set of samples, where each sample
+//   references a sequence of locations, and where each location belongs
+//   to a mapping.
+// - There is a N->1 relationship from sample.location_id entries to
+//   locations. For every sample.location_id entry there must be a
+//   unique Location with that index.
+// - There is an optional N->1 relationship from locations to
+//   mappings. For every nonzero Location.mapping_id there must be a
+//   unique Mapping with that index.
+
+syntax = "proto3";
+
+package opentelemetry.proto.profiles.v1experimental;
+
+import "opentelemetry/proto/common/v1/common.proto";
+
+option csharp_namespace = "OpenTelemetry.Proto.Profiles.V1Experimental";
+option go_package = "go.opentelemetry.io/proto/otlp/profiles/v1experimental";
+
+// Represents a complete profile, including sample types, samples,
+// mappings to binaries, locations, functions, string table, and additional metadata.
+message Profile {
+  // A description of the samples associated with each Sample.value.
+  // For a cpu profile this might be:
+  //   [["cpu","nanoseconds"]] or [["wall","seconds"]] or [["syscall","count"]]
+  // For a heap profile, this might be:
+  //   [["allocations","count"], ["space","bytes"]],
+  // If one of the values represents the number of events represented
+  // by the sample, by convention it should be at index 0 and use
+  // sample_type.unit == "count".
+  repeated ValueType sample_type = 1;
+  // The set of samples recorded in this profile.
+  repeated Sample sample = 2;
+  // Mapping from address ranges to the image/binary/library mapped
+  // into that address range.  mapping[0] will be the main binary.
+  repeated Mapping mapping = 3;
+  // Locations referenced by samples via location_indices.
+  repeated Location location = 4;
+  // Array of locations referenced by samples.
+  repeated int64 location_indices = 15;
+  // Functions referenced by locations.
+  repeated Function function = 5;
+  // Lookup table for attributes.
+  repeated opentelemetry.proto.common.v1.KeyValue attribute_table = 16;
+  // Represents a mapping between Attribute Keys and Units.
+  repeated AttributeUnit attribute_units = 17;
```

Comment:
In terms of inter-operability with the existing [pprof](https://github.com/google/pprof/blob/29b9bb013b0f45d835bcffd9a0995e0d77127453/proto/profile.proto) spec, I do think it would be a good idea to offset the additional field IDs by a an offset. I arbitrarily chose `+32`. If in the meantime the upstream pprof adds some fields (up to 15 🙂 ), we can stay interoperable. If we don't have that gap and for some reason additional fields are added, we are in trouble.


```suggestion
  // Array of locations referenced by samples.
  repeated int64 location_indices = 33;
  // Functions referenced by locations.
  repeated Function function = 5;
  // Lookup table for attributes.
  repeated opentelemetry.proto.common.v1.KeyValue attribute_table = 34;
  // Represents a mapping between Attribute Keys and Units.
  repeated AttributeUnit attribute_units = 35;
```

Reply:
Ideally we also make sure they are not getting accidentally used by defining them as reserved

```
  // Those field ids are reserved for future use by the upstream pprof definition
  reserved 15 to 32;
```

Reply:
Good point. Collisions like this would be really hard to handle in the future so I think it is worth trying to address this now.

One slight improvement I would suggest is that we could make it so that all new fields on all structs start from 50, this way it's a little easier to distinguish and gives us a little more room.

Reply:
I'm worried this would solve the wrong problem and will make our payloads bigger.

IMO we need to either make sure that we DO NOT become a fork of pprof OR give up on pprof compatibility completely. Otherwise we get all the legacy of pprof + all the complexity of trying to stay compatible with it, which means we're getting the worst of both worlds.

Protobuf field tags use 3 bits for the wire type, leaving us with 4 bits before we exceed the threshold for going from a 1-byte to a 2-byte varint for the tag. 4 bits are means that the field numbers `0-15` can be encoded efficiently, but `16+` will cause the tags to become twice as big. For fields inside of message types that are used infrequently this won't matter, but I could imagine it to be significant for message types like `Sample` that are used frequently. See [protobuf wire spec](https://protobuf.dev/programming-guides/encoding/) for more details.

Reply:
When was the last time we spoke with pprof team? I think this PR is a significant milestone and we should sync with them. Last info I have is they were willing to entertain the idea that pprof's canonical Protobuf definitions live at Otel. 
@jsuereth can we set up a call and see if we can make more progress on this?

Reply:
@felixge fair point on the size. I forgot that wire type goes into the same varint, and thought we had 7 bits to work with.